File: project-storage-api/src/main/kotlin/mike/dev/project_api/controller/ProjectController.kt
package mike.dev.project_api.controller

import mike.dev.project_api.dto.*
import mike.dev.project_api.service.ProjectService
import jakarta.validation.Valid
import mike.dev.project_api.model.Channel
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/projects")
@CrossOrigin(
    origins = ["http://localhost:3000", "http://localhost:3003"],
    allowedHeaders = ["Authorization", "Content-Type", "X-User-Id"],
    exposedHeaders = ["X-User-Id"]
)
class ProjectController(
    private val projectService: ProjectService
) {
    @PostMapping("/create")
    @ResponseStatus(HttpStatus.CREATED)
    fun createProject(
        @RequestBody request: CreateProjectRequest,
        @RequestHeader("X-User-Id") userId: String
    ): ProjectDto {
        return projectService.createProject(request, userId)
    }

    @GetMapping("/{id}")
    fun getProject(
        @PathVariable id: Long
    ): ProjectDto {
        return projectService.getProject(id)
    }

    @GetMapping
    fun getProjectsByUser(
        @RequestHeader("X-User-Id") userId: String
    ): List<ProjectDto> {
        return projectService.getProjectsByUser(userId)
    }

    @PutMapping("/{id}")
    fun updateProject(
        @PathVariable id: Long,
        @RequestBody @Valid request: UpdateProjectRequest
    ): ProjectDto {
        return projectService.updateProject(id, request)
    }

    @DeleteMapping("/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun deleteProject(@PathVariable id: Long) {
        projectService.deleteProject(id)
    }

    @PostMapping("/{projectId}/channels")
    fun addChannel(
        @PathVariable projectId: Long,
        @RequestBody request: Any
    ): ProjectDto {
        return projectService.addChannel(projectId, request as CreateChannelRequest)
    }

    @PutMapping("/{projectId}/channels/{channelId}")
    fun updateChannel(
        @PathVariable projectId: Long,
        @PathVariable channelId: Long,
        @RequestBody @Valid request: UpdateChannelRequest
    ): ResponseEntity<Channel?> {
        val updatedChannel = projectService.updateChannel(
            projectId, channelId, request
        )
        return ResponseEntity.ok<Channel>(updatedChannel)
    }

    @DeleteMapping("/{projectId}/channels/{channelId}")
    fun deleteChannel(
        @PathVariable projectId: Long,
        @PathVariable channelId: Long
    ): ProjectDto {
        return projectService.deleteChannel(projectId, channelId)
    }

    @PostMapping("/{projectId}/channels/{channelId}/tracks")
    fun createTrack(
        @PathVariable projectId: Long,
        @PathVariable channelId: Long,
        @RequestBody request: CreateTrackRequest
    ): TrackDto {
        return projectService.createTrack(projectId, channelId, request)
    }

    @PutMapping("/{projectId}/channels/{channelId}/tracks/{trackId}")
    fun updateTrack(
        @PathVariable projectId: Long,
        @PathVariable channelId: Long,
        @PathVariable trackId: Long,
        @RequestBody request: UpdateTrackRequest
    ): TrackDto {
        return projectService.updateTrack(projectId, channelId, trackId, request)
    }

    @DeleteMapping("/{projectId}/channels/{channelId}/tracks/{trackId}")
    fun deleteTrack(
        @PathVariable projectId: Long,
        @PathVariable channelId: Long,
        @PathVariable trackId: Long
    ): ResponseEntity<Unit> {
        projectService.deleteTrack(projectId, channelId, trackId)
        return ResponseEntity.ok().build()
    }
} 

File: projects-api-node/src/controllers/ProjectController.ts
import { Router } from 'express';
import { ProjectService } from '../services/ProjectService.js';
import { CreateProjectRequest, UpdateProjectRequest } from '../dtos/ProjectDto.js';
import { CreateChannelRequest, UpdateChannelRequest } from '../dtos/ChannelDto.js';
import { CreateTrackRequest, UpdateTrackRequest } from '../dtos/TrackDto.js';
import { validateRequest } from '../middleware/validateRequest.js';

export class ProjectController {
    public router: Router;
    private projectService: ProjectService;

    constructor() {
        this.router = Router();
        this.projectService = new ProjectService();
        this.initializeRoutes();
    }

    private initializeRoutes(): void {
        // Project routes
        this.router.post('/create', validateRequest(CreateProjectRequest), this.createProject.bind(this));
        this.router.get('/:id', this.getProject.bind(this));
        this.router.get('/', this.getProjectsByUser.bind(this));
        this.router.put('/:id', validateRequest(UpdateProjectRequest), this.updateProject.bind(this));
        this.router.delete('/:id', this.deleteProject.bind(this));

        // Channel routes
        this.router.post('/:projectId/channels', validateRequest(CreateChannelRequest), this.addChannel.bind(this));
        this.router.put('/:projectId/channels/:channelId', validateRequest(UpdateChannelRequest), this.updateChannel.bind(this));
        this.router.delete('/:projectId/channels/:channelId', this.deleteChannel.bind(this));

        // Track routes
        this.router.post('/:projectId/channels/:channelId/tracks', validateRequest(CreateTrackRequest), this.createTrack.bind(this));
        this.router.put('/:projectId/channels/:channelId/tracks/:trackId', validateRequest(UpdateTrackRequest), this.updateTrack.bind(this));
        this.router.delete('/:projectId/channels/:channelId/tracks/:trackId', this.deleteTrack.bind(this));
    }

    public async createProject(req: any, res: any): Promise<void> {
        try
        {
            console.log('ProjectController - Creating project with body:', JSON.stringify(req.body, null, 2));
            console.log('ProjectController - User ID from headers:', req.headers['x-user-id']);
            const project = await this.projectService.createProject(req.body, req.headers['x-user-id']);
            console.log('ProjectController - Project created successfully:', JSON.stringify(project, null, 2));
            res.status(201).json(project);
        } catch (error: any)
        {
            console.error('ProjectController - Error creating project:', error);
            res.status(500).json({ error: error.message });
        }
    }

    public async getProject(req: any, res: any): Promise<void> {
        try
        {
            const project = await this.projectService.getProject(parseInt(req.params.id));
            res.json(project);
        } catch (error: any)
        {
            res.status(500).json({ error: error.message });
        }
    }

    public async getProjectsByUser(req: any, res: any): Promise<void> {
        try
        {
            const projects = await this.projectService.getProjectsByUser(req.headers['x-user-id']);
            res.json(projects);
        } catch (error: any)
        {
            res.status(500).json({ error: error.message });
        }
    }

    public async updateProject(req: any, res: any): Promise<void> {
        try
        {
            const project = await this.projectService.updateProject(parseInt(req.params.id), req.body);
            res.json(project);
        } catch (error: any)
        {
            res.status(500).json({ error: error.message });
        }
    }

    public async deleteProject(req: any, res: any): Promise<void> {
        try
        {
            await this.projectService.deleteProject(parseInt(req.params.id));
            res.status(204).send();
        } catch (error: any)
        {
            res.status(500).json({ error: error.message });
        }
    }

    public async addChannel(req: any, res: any): Promise<void> {
        try
        {
            const project = await this.projectService.addChannel(parseInt(req.params.projectId), req.body);
            res.json(project);
        } catch (error: any)
        {
            res.status(500).json({ error: error.message });
        }
    }

    public async updateChannel(req: any, res: any): Promise<void> {
        try
        {
            const project = await this.projectService.updateChannel(
                parseInt(req.params.projectId),
                parseInt(req.params.channelId),
                req.body
            );
            res.json(project);
        } catch (error: any)
        {
            res.status(500).json({ error: error.message });
        }
    }

    public async deleteChannel(req: any, res: any): Promise<void> {
        try
        {
            const project = await this.projectService.deleteChannel(
                parseInt(req.params.projectId),
                parseInt(req.params.channelId)
            );
            res.json(project);
        } catch (error: any)
        {
            res.status(500).json({ error: error.message });
        }
    }

    public async createTrack(req: any, res: any): Promise<void> {
        try
        {
            const project = await this.projectService.createTrack(
                parseInt(req.params.projectId),
                parseInt(req.params.channelId),
                req.body
            );
            res.json(project);
        } catch (error: any)
        {
            res.status(500).json({ error: error.message });
        }
    }

    public async updateTrack(req: any, res: any): Promise<void> {
        try
        {
            const project = await this.projectService.updateTrack(
                parseInt(req.params.projectId),
                parseInt(req.params.channelId),
                parseInt(req.params.trackId),
                req.body
            );
            res.json(project);
        } catch (error: any)
        {
            res.status(500).json({ error: error.message });
        }
    }

    public async deleteTrack(req: any, res: any): Promise<void> {
        try
        {
            const project = await this.projectService.deleteTrack(
                parseInt(req.params.projectId),
                parseInt(req.params.channelId),
                parseInt(req.params.trackId)
            );
            res.json(project);
        } catch (error: any)
        {
            res.status(500).json({ error: error.message });
        }
    }
} 